modules/ROOT/pages/constraints/managing-constraints.adoc

```diff
@@ -0,0 +1,1598 @@
+
+.Create a relationship that violates an existing relationship property type constraint
+======
+
+
+.Create a `PART_OF` relationship with a `STRING` `order` property, given a property type constraint on the relationship type `PART_OF` restricting the `order` property to `INTEGER` values
+[source, cypher, role=test-fail]
+----
+MATCH (movie:Movie {title:'Iron Man'}), (franchise:Franchise {name:'MCU'})
+CREATE (movie)-[part:PART_OF {order: '1'}]->(franchise)
+----
+
+In this case, the relationship is not created because the `order` property is in conflict with an existing constraint.
+
+.Error message
+[source, error]
+----
+Relationship(0) with type `PART_OF` has property `order` of wrong type `String`. Allowed types: INTEGER
+----
+
+======
+
+
+.Create a node that violates existing node property key constraint
+======
+
+.Create an `Actor` node without a `firstname` property, given a node key constraint on `:Actor(firstname, surname)`
+[source, cypher, role=test-fail]
+----
+CREATE (actor:Actor {surname: 'Wood'})
+----
+
+In this case, the node is not created because it is missing the `firstname` property which is in conflict with an existing constraint.
+
+.Error message
+[source, error]
+----
+Node(0) with label `Actor` must have the properties (`firstname`, `surname`)
+----
+
+======
+
+[[fail-to-create-constraint-due-to-existing-data]]
+==== Creating constraints when there exists conflicting data will fail
+
+This applies to all constraint types.
+
+.Create a node property uniqueness constraint when conflicting nodes exist
+======
+
+
+////
+[source, cypher, role=test-setup]
+----
+CREATE (book:Book {isbn: '9780393972832', title: 'Moby Dick'});
+CREATE (book:Book {isbn: '9780763630188', title: 'Moby Dick'})
+----
+////
+
+.Create a constraint requiring `Book` nodes to have unique `title` properties, when there already exists two nodes with the same `title`
+[source, cypher, role=test-fail]
+----
+CREATE CONSTRAINT book_title FOR (book:Book) REQUIRE book.title IS UNIQUE
+----
+
+In this case, the constraint cannot be created because it is in conflict with the existing graph.
+Either use xref:indexes/search-performance-indexes/managing-indexes.adoc[indexes] instead, or remove/correct the offending nodes and then re-apply the constraint.
+
+.Error message
+[source, error]
+----
+Unable to create Constraint( name='book_title', type='UNIQUENESS', schema=(:Book {title}) ):
+Both Node(0) and Node(1) have the label `Book` and property `title` = 'Moby Dick'
+----
+
+The constraint creation fails on the first offending nodes that are found.
+This does not guarantee that there are no other offending nodes in the graph.
+Therefore, all the data should be checked and cleaned up before re-attempting the constraint creation.
+
+.Find all offending nodes with the non-unique property values for the constraint above
+[source, cypher]
+----
+MATCH (book1:Book), (book2:Book)
+WHERE book1.title = book2.title AND NOT book1 = book2
+RETURN book1, book2
+----
```

Comment:
Since we only cover two of the four constraint type we are missing out on some of these helpful matches now.

Maybe add after both examples just the relevant match clauses for either all or at least the remaining constraint types as well (or if you make a table of why they fall above, add it as another column up there)

Reply:
(especially since the key constraint one is missing and its the most complicated one as well)

Reply:
generic versions of all matches:

node property uniqueness match:
```
MATCH (n1:Label), (n2:Label)
WHERE n1.prop = n2.prop AND NOT n1 = n2
RETURN n1, n2
```
rel property uniqueness match:
```
MATCH ()-[r1:REL_TYPE]->(), ()-[r2:REL_TYPE]->()
WHERE r1.prop = r2.prop AND NOT r1 = r2
RETURN r1, r2
```

node property existence match:
```
MATCH (n:Label)
WHERE n.prop IS NULL
RETURN n
```
rel property existence match:
```
MATCH ()-[r:REL_TYPE]-()
WHERE r.prop IS NULL
RETURN r
```

node property type match:
```
MATCH (n:Label)
WHERE n.prop IS NOT :: <TYPE>
RETURN n
```
rel property type match:
```
MATCH ()-[r:REL_TYPE]-()
WHERE r.prop IS NOT :: <TYPE>
RETURN r
```

node key match:
```
MATCH (n1:Label), (n2:Label)
WHERE n1.prop = n2.prop AND NOT n1 = n2
UNWIND [n1, n2] AS node
RETURN node, 'non-unique' AS reason
UNION
MATCH (n:Label)
WHERE n.prop IS NULL
RETURN n AS node, 'non-existing' AS reason
```
rel key match:
```
MATCH ()-[r1:REL_TYPE]->(), ()-[r2:REL_TYPE]->()
WHERE r1.prop = r2.prop AND NOT r1 = r2
UNWIND [r1, r2] AS relationship
RETURN relationship, 'non-unique' AS reason
UNION
MATCH ()-[r:REL_TYPE]-()
WHERE r.prop IS NULL
RETURN r AS relationship, 'non-existing' AS reason
```